File: Semestrul 4/MPP/lab08/TriathlonPersistence/Impl/RefereeDisciplineRepositoryImpl.cs
using System;
using System.Collections.Generic;
using System.Data.SQLite;
using Triathlon.Model;
using Triathlon.Persistence.Abstracts;
using Triathlon.Persistence.Interfaces;

namespace Triathlon.Persistence.Impl
{
    public class RefereeDisciplineRepositoryImpl : AbstractDbRepository<long, RefereeDiscipline>, IRefereeDisciplineRepository
    {
        private readonly IRefereeRepository _refereeRepository;
        private readonly IEventRepository _eventRepository;

        public RefereeDisciplineRepositoryImpl(IRefereeRepository refereeRepository,
                                              IEventRepository eventRepository)
            : base("RefereeDisciplines", "id")
        {
            _refereeRepository = refereeRepository;
            _eventRepository = eventRepository;
        }

        protected override string GetInsertParameters()
        {
            return "@RefereeId, @EventId, @DisciplineId, @AssignedDate";
        }

        protected override string GetUpdateParameters()
        {
            return "referee_id=@RefereeId, event_id=@EventId, discipline_id=@DisciplineId, assigned_date=@AssignedDate";
        }

        protected override int GetUpdateParameterCount()
        {
            return 4;
        }

        protected override RefereeDiscipline CreateEntity(SQLiteDataReader data)
        {
            long id = data.GetInt64(data.GetOrdinal("id"));
            long refereeId = data.GetInt64(data.GetOrdinal("referee_id"));
            long eventId = data.GetInt64(data.GetOrdinal("event_id"));
            int disciplineId = data.GetInt32(data.GetOrdinal("discipline_id"));

            Referee referee = _refereeRepository.GetById(refereeId);
            Event @event = _eventRepository.GetById(eventId);
            Discipline discipline = disciplineId switch
            {
                1 => Discipline.RUNNING,
                2 => Discipline.SWIMMING,
                3 => Discipline.CYCLING,
                _ => throw new InvalidOperationException($"Unknown discipline ID: {disciplineId}")
            };

            return new RefereeDiscipline(id, referee, @event, discipline);
        }

        protected override void SaveEntity(SQLiteCommand command, RefereeDiscipline entity)
        {
            command.Parameters.AddWithValue("@RefereeId", entity.Referee.Id);
            command.Parameters.AddWithValue("@EventId", entity.Event.Id);
            command.Parameters.AddWithValue("@DisciplineId", (int)entity.Discipline);
            command.Parameters.AddWithValue("@AssignedDate", DateTime.Now);         }

        protected override long GetEntityId(RefereeDiscipline entity)
        {
            return entity.Id;
        }

        public RefereeDiscipline FindByRefereeId(long refereeId)
        {
            string sql = $"SELECT * FROM {TableName} WHERE referee_id = @RefereeId";
            List<RefereeDiscipline> refereeDisciplines = new List<RefereeDiscipline>();

            try
            {
                using (SQLiteCommand command = new SQLiteCommand(sql, Connection))
                {
                    if (Connection.State != System.Data.ConnectionState.Open)
                        Connection.Open();

                    command.Parameters.AddWithValue("@RefereeId", refereeId);

                    using (SQLiteDataReader reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            refereeDisciplines.Add(CreateEntity(reader));
                        }
                    }
                }

                return refereeDisciplines.Count > 0 ? refereeDisciplines[0] : null;
            }
            catch (Exception ex)
            {
                throw new Exception($"Error finding referee discipline by referee ID: {ex.Message}", ex);
            }
        }

        public IEnumerable<RefereeDiscipline> FindByEventId(long eventId)
        {
            string sql = $"SELECT * FROM {TableName} WHERE event_id = @EventId";
            List<RefereeDiscipline> refereeDisciplines = new List<RefereeDiscipline>();

            try
            {
                using (SQLiteCommand command = new SQLiteCommand(sql, Connection))
                {
                    if (Connection.State != System.Data.ConnectionState.Open)
                        Connection.Open();

                    command.Parameters.AddWithValue("@EventId", eventId);

                    using (SQLiteDataReader reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            refereeDisciplines.Add(CreateEntity(reader));
                        }
                    }
                }

                return refereeDisciplines;
            }
            catch (Exception ex)
            {
                throw new Exception($"Error finding referee disciplines by event ID: {ex.Message}", ex);
            }
        }

        public IEnumerable<RefereeDiscipline> FindByDisciplineId(long disciplineId)
        {
            string sql = $"SELECT * FROM {TableName} WHERE discipline_id = @DisciplineId";
            List<RefereeDiscipline> refereeDisciplines = new List<RefereeDiscipline>();

            try
            {
                using (SQLiteCommand command = new SQLiteCommand(sql, Connection))
                {
                    if (Connection.State != System.Data.ConnectionState.Open)
                        Connection.Open();

                    command.Parameters.AddWithValue("@DisciplineId", disciplineId);

                    using (SQLiteDataReader reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            refereeDisciplines.Add(CreateEntity(reader));
                        }
                    }
                }

                return refereeDisciplines;
            }
            catch (Exception ex)
            {
                throw new Exception($"Error finding referee disciplines by discipline ID: {ex.Message}", ex);
            }
        }

        public IEnumerable<RefereeDiscipline> FindByRefereeAndEvent(long refereeId, long eventId)
        {
            string sql = $"SELECT * FROM {TableName} WHERE referee_id = @RefereeId AND event_id = @EventId";
            List<RefereeDiscipline> refereeDisciplines = new List<RefereeDiscipline>();

            try
            {
                using (SQLiteCommand command = new SQLiteCommand(sql, Connection))
                {
                    if (Connection.State != System.Data.ConnectionState.Open)
                        Connection.Open();

                    command.Parameters.AddWithValue("@RefereeId", refereeId);
                    command.Parameters.AddWithValue("@EventId", eventId);

                    using (SQLiteDataReader reader = command.ExecuteReader())
                    {
                        while (reader.Read())
                        {
                            refereeDisciplines.Add(CreateEntity(reader));
                        }
                    }
                }

                return refereeDisciplines;
            }
            catch (Exception ex)
            {
                throw new Exception($"Error finding referee disciplines by referee and event ID: {ex.Message}", ex);
            }
        }

        public RefereeDiscipline FindByRefereeEventAndDiscipline(long refereeId, long eventId, long disciplineId)
        {
            string sql = $"SELECT * FROM {TableName} WHERE referee_id = @RefereeId AND event_id = @EventId AND discipline_id = @DisciplineId";

            try
            {
                using (SQLiteCommand command = new SQLiteCommand(sql, Connection))
                {
                    if (Connection.State != System.Data.ConnectionState.Open)
                        Connection.Open();

                    command.Parameters.AddWithValue("@RefereeId", refereeId);
                    command.Parameters.AddWithValue("@EventId", eventId);
                    command.Parameters.AddWithValue("@DisciplineId", disciplineId);

                    using (SQLiteDataReader reader = command.ExecuteReader())
                    {
                        if (reader.Read())
                        {
                            return CreateEntity(reader);
                        }
                    }
                }

                return null;
            }
            catch (Exception ex)
            {
                throw new Exception($"Error finding referee discipline by referee, event, and discipline ID: {ex.Message}", ex);
            }
        }
    }
}

File: Semestrul 4/SDI/lab12/java/TriathlonPersistence/src/main/java/triathlon/persistence/impl/RefereeDisciplineRepositoryImpl.java
package triathlon.persistence.impl;

import triathlon.model.*;
import triathlon.persistence.interfaces.*;
import triathlon.persistence.abstracts.AbstractDBRepository;

import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;
import java.util.Optional;

public class RefereeDisciplineRepositoryImpl extends AbstractDBRepository<Long, RefereeDiscipline> implements RefereeDisciplineRepository {
    private final RefereeRepository refereeRepository;
    private final EventRepository eventRepository;

    public RefereeDisciplineRepositoryImpl(RefereeRepository refereeRepository,
                                           EventRepository eventRepository) {
        super("RefereeDisciplines", "id");
        this.refereeRepository = refereeRepository;
        this.eventRepository = eventRepository;
    }

    @Override
    protected String getInsertParameters() {
        return "referee_id, event_id, discipline_id, assigned_date";
    }

    @Override
    protected String getUpdateParameters() {
        return "referee_id=?, event_id=?, discipline_id=?, assigned_date=?";
    }

    @Override
    protected int getUpdateParameterCount() {
        return 4;
    }

    @Override
    protected String getColumns() {
        return "";
    }

    @Override
    protected RefereeDiscipline createEntity(ResultSet data) throws SQLException {
        Long id = data.getLong("id");
        Long refereeId = data.getLong("referee_id");
        Long eventId = data.getLong("event_id");
        int disciplineId = data.getInt("discipline_id");

        Referee referee = refereeRepository.findOne(refereeId).orElse(null);
        Event event = eventRepository.findOne(eventId).orElse(null);
        Discipline discipline = switch (disciplineId) {
            case 1 -> Discipline.RUNNING;
            case 2 -> Discipline.SWIMMING;
            case 3 -> Discipline.CYCLING;
            default -> throw new SQLException("Unknown discipline ID: " + disciplineId);
        };

        RefereeDiscipline refDiscipline = new RefereeDiscipline(id, referee, event, discipline);

        return refDiscipline;
    }

    @Override
    protected void saveEntity(PreparedStatement preparedStatement, RefereeDiscipline entity) throws SQLException {
        preparedStatement.setLong(1, entity.getReferee().getId());
        preparedStatement.setLong(2, entity.getEvent().getId());
        preparedStatement.setInt(3, entity.getDiscipline().ordinal());
    }

    @Override
    public Optional<RefereeDiscipline> findByRefereeId(Long refereeId) {
        String sql = "SELECT * FROM " + tableName + " WHERE referee_id = ?";
        List<RefereeDiscipline> refereeDisciplines = new ArrayList<>();

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setLong(1, refereeId);

            try (ResultSet resultSet = statement.executeQuery()) {
                while (resultSet.next()) {
                    refereeDisciplines.add(createEntity(resultSet));
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return Optional.of(refereeDisciplines.getFirst());
    }

    @Override
    public List<RefereeDiscipline> findByEventId(Long eventId) {
        String sql = "SELECT * FROM " + tableName + " WHERE event_id = ?";
        List<RefereeDiscipline> refereeDisciplines = new ArrayList<>();

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setLong(1, eventId);

            try (ResultSet resultSet = statement.executeQuery()) {
                while (resultSet.next()) {
                    refereeDisciplines.add(createEntity(resultSet));
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return refereeDisciplines;
    }

    @Override
    public List<RefereeDiscipline> findByDisciplineId(Long disciplineId) {
        String sql = "SELECT * FROM " + tableName +" WHERE discipline_id = ?";
        List<RefereeDiscipline> refereeDisciplines = new ArrayList<>();

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setLong(1, disciplineId);

            try (ResultSet resultSet = statement.executeQuery()) {
                while (resultSet.next()) {
                    refereeDisciplines.add(createEntity(resultSet));
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return refereeDisciplines;
    }

    @Override
    public List<RefereeDiscipline> findByRefereeAndEvent(Long refereeId, Long eventId) {
        String sql = "SELECT * FROM " + tableName + " WHERE referee_id = ? AND event_id = ?";
        List<RefereeDiscipline> refereeDisciplines = new ArrayList<>();

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setLong(1, refereeId);
            statement.setLong(2, eventId);

            try (ResultSet resultSet = statement.executeQuery()) {
                while (resultSet.next()) {
                    refereeDisciplines.add(createEntity(resultSet));
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return refereeDisciplines;
    }

    @Override
    public RefereeDiscipline findByRefereeEventAndDiscipline(Long refereeId, Long eventId, Long disciplineId) {
        String sql = "SELECT * FROM " + tableName + " WHERE referee_id = ? AND event_id = ? AND discipline_id = ?";

        try (PreparedStatement statement = connection.prepareStatement(sql)) {
            statement.setLong(1, refereeId);
            statement.setLong(2, eventId);
            statement.setLong(3, disciplineId);

            try (ResultSet resultSet = statement.executeQuery()) {
                if (resultSet.next()) {
                    return createEntity(resultSet);
                }
            }
        } catch (SQLException e) {
            e.printStackTrace();
        }

        return null;
    }
}